### ISSUE TITLE
Support HLS through src= on iOS

### ISSUE THREAD

**Issue by joeyparrish** [google/shaka-player#997]
Related to #816, we will be supporting HLS through Apple's built-in HLS implementation on iOS.  This depends on the refactoring in #816, plus additional checks for Apple's built-in HLS implementation.

As in #816, using Apple's HLS implementation on iOS means most of Shaka Player's configuration will have no effect. The browser will be responsible for streaming, buffering, etc. Side-loaded captions would also fail (because we would not be streaming text, either). Shaka Player would only be able to control DRM and license requests.

But this does mean some playback on iOS with the same player you use on other browsers.

**Comment by chrisfillmore** [comment]
Is the support for `src=` on iOS substantially different than on desktop Safari?

**Comment by avelad** [comment]
Will there be support for Fairplay on iOS with Key System "com.apple.fps.1_0"? This is probably related to https://github.com/google/shaka-player/issues/382

**Comment by joeyparrish** [comment]
@chrisfillmore, no, it will be much the same.  The difference is that desktop Safari has MediaSource, so desktop Safari will be able to play HLS through either mechanism.

@avelad, hopefully this simplifies FairPlay for both iOS and desktop Safari.

**Comment by chrisfillmore** [comment]
I'd like to offer a dissenting opinion. I invite others to chime in if they disagree, particularly if this enhancement would add value for them.

A few months ago we attempted an integration with another vendor's HTML5 player, and decided to drop it after several weeks of work. One aspect of this player which was unappealing was how it attempted to be a universal solution for all use cases. We were looking to solve a particular problem, and were burdened with solutions to problems we didn't have.

Something that is appealing about Shaka is how it is relatively lightweight. It solves the particular _difficult_ problem of adaptive streaming. Setting `video.src=` is not a difficult problem to solve (in the context of streaming video). Apple even provides a demo in the FPS SDK, if you need support for Safari.

**Comment by joeyparrish** [comment]
@chrisfillmore, thank you for your feedback.  I will try to solicit broader feedback from the community before proceeding.

**Comment by radiantmediaplayer** [comment]
We (happily) use Shaka player in our player and started recently to support fmp4-HLS with it. We have a complete different player for iOS using native src though, notably due to the fact that video ads work differently on iOS (e.g. src needs to be changed on each ad break). However if you can crack FPS on iOS Safari we would be more than happy to test Shaka player on that aspect.

**Comment by TigerHuang** [comment]
iOS 11.3 (Safari 11.1) support EME
https://developer.apple.com/library/content/releasenotes/General/WhatsNewInSafari/Articles/Safari_11_1.html

**Comment by joeyparrish** [comment]
@chrisfillmore, I appreciate your feedback and I acknowledge your dissenting opinion on this.  I wanted to tie up one loose end now that we've started work:

> If support for src= increases complexity in Shaka to the extent that it makes it more difficult to develop solutions for streaming, then IMO it seems like more of a burden than a help. I would argue that client applications should not have a hard dependency on Shaka -- whatever abstraction sits between the client and Shaka should allow the app to play via src=.

What we've found so far is that the work we've started to implement src= support is actually cleaning up a lot of the most complex and difficult to maintain parts of the `Player` class.  So rather than increasing the complexity of Shaka Player, I believe this work is significantly improving Shaka Player in the long run.  It's a lot of work, but it seems worthwhile.

Apps are, of course, still free to use src= for themselves, and will never be required to depend on Shaka more than they want to.
